File: cpp/linked-list.cpp
#include <iostream>

using namespace std;

//Create new node structure
struct node{
    int data;
    node* next;
};
//create Head pointer to point to the first node
node* head = NULL;

//Functions declaration
void insertNode(int value);
void displayLinkedList();
void deleteNode(int value);
void insertBeginning(int value);
void deleteBeginning();
void deleteEnd();


int main()
{
    //iNSERTION
    insertNode(5);
    insertNode(10);
    insertNode(15);
    insertNode(20);
    insertNode(25);
    //CALL FUNCTIONS
    displayLinkedList();
    deleteNode(15);
    displayLinkedList();
    insertBeginning(15);
    displayLinkedList();
    deleteBeginning();
    displayLinkedList();
    deleteEnd();
    displayLinkedList();

    return 0;
}
void insertNode(int value){

    node* new_node;
    node* last;
    new_node = new node;
    new_node->data = value;
    //Head isn't connected to any linkedlist
    if(head == NULL){
        head = new_node;
        new_node->next = NULL;
    //Here head point to the first node
    }else{
        last = head;
        while(last->next != NULL){
            last = last->next;
        }
        last->next = new_node;
        new_node->next = NULL;
    }
}

void displayLinkedList(){
    node* displayData;
    if(head == NULL){
        cout<<"This linked list is empty";
    }
    else{
        displayData = head;
        cout<<displayData->data;
        while(displayData->next != NULL){
            displayData = displayData->next;
            cout<<"\t"<<displayData->data;
        }
    }
    cout<<endl;
}

void deleteNode(int value){
    node* current_node, * prev_node;
    current_node = head;
    prev_node = head;

    if(current_node->data == value){
        head = current_node->next;
        free(current_node);
        return;
    }
    while(current_node->data != value){
        prev_node = current_node;
        current_node = current_node->next;
    }
    prev_node->next = current_node->next;
    free(current_node);
}

void insertBeginning(int value){
    node* new_node = new node;
    new_node->data = value;
    new_node->next = head;
    head = new_node;
}

void deleteBeginning(){
    node* first_node;
    if(head == NULL){
        cout<<"Empty linked list";
    }
    else{
        first_node = head;
        head = first_node->next;
        free(first_node);
    }
}

void deleteEnd(){
    if (head == NULL){
        cout<<"Linked List is already empty";
    }
    else if (head->next == NULL){
        delete(head);
        head = NULL;
    }
    else{
        node* ptr = head;
        while(ptr->next->next != NULL){
            ptr = ptr->next;
        }
        delete(ptr->next);
        ptr->next = NULL;
    }
}

File: javaScript/linked-list.js
//Structure of any node 
class Node {
    constructor(data , next=null){
        this.data = data
        this.next = next
    }
}

//Linked list class
class linkedList{
    constructor(){      //We consider that the linked list is empty at the beginning
        this.head = null
        this.size = 0
    }
    //Insert first node
    firstInsert(data){
        this.head = new Node(data, this.head)
        this.size++
    }
    //Insert last node
    lastinsert(data){
        let node = new Node(data)
        if(!this.head){
            this.head = node;
            this.size++
        }else{
            let pointer = this.head
            while(pointer.next){
                pointer = pointer.next
            }
            pointer.next = node
            this.size++
        }
    }
    //insert at any position 
    indexInsert(data, index){
        if(index<0 || index>this.size){
            console.log('No negtive index are allowed')
            return
        }
        if(index == 0 ){
            this.head = new Node(data, this.head)
            this.size++
            return 
        }
        const node = new Node(data)
        let currentNode, prevNode
        currentNode = this.head
        let counter = 0 
        //to get the previous node index 
        while (counter < index){
            prevNode = currentNode
            counter++
            currentNode = currentNode.next
        }
        prevNode.next = node
        node.next = currentNode
        this.size++
    }
    //get specific node by index
    getByIndex(index){
        let currentNode;
        currentNode = this.head
        let counter = 0
        while(currentNode){
            if(counter == index){
                console.log(index+'\'s value is :'+currentNode.data)
                return
            }
            else{
                counter++
                currentNode = currentNode.next
            }
        }
    }
    //remove specific node
    removeAt(index){
        if(index<0 || index>=this.size){
            return
        }
        let currentNode, prevNode
        currentNode = this.head
        let counter = 0
        if(index === 0){
            this.head = currentNode.next
        }else{
            while(counter < index){
                counter++
                prevNode = currentNode
                currentNode = currentNode.next
            }
            currentNode = currentNode.next
            prevNode.next = currentNode
        }
    }
    //clear all the linked list
    clear(){
        this.head = 0
        this.size = 0
    }
    //Display Linked list data
    display(){
        let data = this.head
        while(data){
            console.log(data.data)
            data = data.next
        }
        console.log('linked list size is :' +  this.size)
    }
}


//Try part
const list = new linkedList()
list.firstInsert(100)
list.firstInsert(200)
list.firstInsert(300)
list.lastinsert(400)
list.lastinsert(500)
list.indexInsert(150,3)
const list2 = new linkedList()
list2.firstInsert(100)
list2.lastinsert(200)
list2.lastinsert(300)
list2.removeAt(0)
//print linked list data
list2.display()



